journal/history.go

```diff
@@ -0,0 +1,63 @@
+package journal
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+)
+
+var mux = sync.Mutex{}
+
+// Entry contains the command-line used to invoke the process associated with this item
+// and its journal ID.
+type Entry struct {
+	ID, Description string
+	Start           time.Time
+}
+
+// A list of commands that have been invoked
+var history = make([]Entry, 0)
+
+// Lookup to identify whether a journal ID has already been recorded in history. There's a small
+// chance the journal ID may not be unique in history. We handle that by ignoring any entries we
+// find that are more than an hour old.
+var jidToHistory = make(map[string]time.Time)
+
+// registerCommand appends the command description to history if it hasn't been registered before.
+func (e Entry) registerCommand() {
+	// Skip if expiration is still in the future.
+	if expire, ok := jidToHistory[e.ID]; ok && expire.After(time.Now()) {
+		return
+	}
+
+	// Expired or not yet set. Lock to make sure we only register the command once.
+	mux.Lock()
+	defer mux.Unlock()
+
+	// Check again in case something else beat us to it.
+	if expire, ok := jidToHistory[e.ID]; ok && expire.After(time.Now()) {
+		return
+	}
+
+	// Set to expire in an hour.
+	jidToHistory[e.ID] = time.Now().Add(1 * time.Hour)
+	if e.Start.IsZero() {
+		e.Start = time.Now()
+	}
+	history = append(history, e)
+	return
+}
+
+// History returns the entire history.
+func History() []Entry {
+	// We only ever add to history, so the slice wont be modified by other operations later.
+	return history
+}
+
+// Open returns a reader to read the journal specified by its ID.
+func Open(journalID string) (io.ReadCloser, error) {
```

Comment:
Is thread-safety a concern here? For example, would it be bad if we open a journal that's also being written to?

Reply:
There shouldn't be. Writing will happen at the end. Worst-case we get a partially written line at the end.